### PR TITLE
add binary search in c

### LINKED ISSUE #44: [Algo] Binary Search [C]

### PR DESCRIPTION
**Pull request by varunkmr19**
**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

Fixes #44 

**By submitting this pull request I confirm I've read and complied with the
below declarations.**

- [x] I have read the [Contribution guidelines](https://github.com/NITSkmOS/Algorithms/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [commit guidelines](https://github.com/NITSkmOS/Algorithms/blob/master/CONTRIBUTING.md#write-good-commit-messages) for this project.
- [x] My code follows the [standard code structure](https://github.com/NITSkmOS/Algorithms/blob/master/CONTRIBUTING.md#samples-code).
- [x] This pull request has a descriptive title. For example, `{Tag}: Add {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous-time span of 7 days.
- [x] This pull request shall only be reviewed and merged once all the checks pass. No maintainer or supporter shall be obliged to review it before this condition is met.
- [x] I have mentioned the issue number correctly (with hyperlink) in this pull request description.

After you submit your pull request, **DO NOT** click the 'Update Branch' button.

### CODE REVIEW

**Comment by varunkmr19** [comment]
I went through the logs of Travis CI. It says there are some extra whitespaces in my source code. So, should I fix it or is it good enough for a merge?

**Comment by sangamcse** [comment]
Yes @varunkmr19, the fix is needed. Also, it would be great if you go through https://github.com/NITSkmOS/Algorithms/blob/master/CONTRIBUTING.md#write-good-commit-messages and do the necessary changes. Also, can you add entry in README.md?

**Comment by sameerchoubey** [comment]
@varunkmr19 There are few things that need to be sorted for this PR to get merged.

a) At line 5, we cannot obtain the size of an array after passing it inside a function because only the first address of the array is passed inside the function. 
When we pass an array to a function, it decays into a pointer to the first element, at which point knowledge of its size is not known.
So, it would be something like this, after the change,
`int main(){`
`int end = sizeof(arr)/sizeof(arr[0]);`
`int ans = binary_search(arr, end, target);`
`}`

b) And do remove the extra spacing.

Everything else looks good.

**Comment by varunkmr19** [comment]
Don't know why It's not passing all the checks. Even after I fixed the code and removed the extra spacing as well.

File: README.md [hunk 1]
```diff
@@ -12,43 +12,43 @@ This repository contains examples of various algorithms written on different pro
...
+| [QuickSort](https://en.wikipedia.org/wiki/Quicksort)                               |   [:octocat:](quicksort/C)    |                                 |   [:octocat:](quicksort/Java)   |     [:octocat:](quicksort/Python)     |
+| [Insertion Sort](https://en.wikipedia.org/wiki/Insertion_sort)                     | [:octocat:](insertion_sort/C) | [:octocat:](insertion_sort/Cpp) |                                 |  [:octocat:](insertion_sort/Python)   |
+| [Counting Sort](https://en.wikipedia.org/wiki/Counting_sort)                       | [:octocat:](counting_sort/C)  | [:octocat:](counting_sort/Cpp)  |                                 |   [:octocat:](counting_sort/Python)   |
+| [Radix Sort](https://en.wikipedia.org/wiki/Radix_sort)                             |                               |   [:octocat:](radix_sort/Cpp)   |                                 |    [:octocat:](radix_sort/Python)     |
+| [Binary Search](https://en.wikipedia.org/wiki/Binary_search_algorithm)             | [:octocat:](binary_search/C)  | [:octocat:](binary_search/Cpp)  |                                 |   [:octocat:](binary_search/Python)   |
```
Comment (sangamcse) [inline README.md:23]: For this file, remove all the changes you did and just keep the below one.
```suggestion
| [Binary Search](https://en.wikipedia.org/wiki/Binary_search_algorithm)                          |                                       | [:octocat:](binary_search/Cpp)        | [:octocat:](binary_search/C)          | [:octocat:](binary_search/Python)     |
```

File: README.md [hunk 2]
```diff
@@ -58,6 +58,7 @@ This repository contains examples of various algorithms written on different pro
 4. Commit your changes: `git commit -am 'Add some feature'`.
 
    Go through [Commit Messages guidelines](CONTRIBUTING.md#write-good-commit-messages)
+
 5. Push to the branch: `git push origin my-new-feature`
 6. Submit a pull request :smile:
```
Comment (sangamcse) [inline README.md:64]: The changes in this file is redundant. Kindly change only line that is required.

**Review by sangamcse — CHANGES_REQUESTED** [review]

**Comment by varunkmr19** [comment]
Finally, I found out what the issue was. I had installed prettier(a VScode extension) which was messing with the alignment.

**Comment by sameerchoubey** [comment]
That's great. But, please squash your commits.

**Comment by sangamcse** [comment]
@varunkmr19 can you please go through the review comments and do the changes that are requested?

File: README.md [hunk 3]
```diff
@@ -12,43 +12,43 @@ This repository contains examples of various algorithms written on different pro
...
+| [Linear Linked List](https://en.wikipedia.org/wiki/Linked_list)        |    [:octocat:](linked_list/C)     |    [:octocat:](linked_list/Cpp)     | [:octocat:](linked_list/Java)  | [:octocat:](linked_list/Python)  |
+| [AVL Tree](https://en.wikipedia.org/wiki/AVL_tree)                     |      [:octocat:](avl_tree/C)      |      [:octocat:](avl_tree/Cpp)      |   [:octocat:](avl_tree/Java)   |   [:octocat:](avl_tree/Python)   |
+| [Binary Search Tree](https://en.wikipedia.org/wiki/Binary_search_tree) |                                   | [:octocat:](binary_search_tree/Cpp) |                                |                                  |
+| [Fenwick Tree](https://en.wikipedia.org/wiki/Fenwick_tree)             |                                   |    [:octocat:](fenwick_tree/Cpp)    | [:octocat:](fenwick_tree/java) | [:octocat:](fenwick_tree/Python) |
+| [Doubly Linked List](https://en.wikipedia.org/wiki/Doubly_linked_list) | [:octocat:](doubly_linked_list/c) |                                     |                                |                                  |
```
Comment (sangamcse) [inline README.md:41]: Please, do not do unnecessary changes
```suggestion
| Data Structure                                                                                  | C                                     | CPP                                   | Java                                  | Python                                |
|:----------------------------------------------------------------------------------------------- |:-------------------------------------:|:-------------------------------------:|:-------------------------------------:|:-------------------------------------:|
| [Queue](https://en.wikipedia.org/wiki/Queue_(abstract_data_type))                               |                                       | [:octocat:](queue/Cpp)                |                                       |                                       |
| [Stack](https://en.wikipedia.org/wiki/Stack_(abstract_data_type))                               | [:octocat:](stack/C )                 |                                       | [:octocat:](stack/Java)               | [:octocat:](stack/Python)             |
| [Linear Linked List](https://en.wikipedia.org/wiki/Linked_list)                                 | [:octocat:](linked_list/C)            | [:octocat:](linked_list/Cpp)          | [:octocat:](linked_list/Java)         | [:octocat:](linked_list/Python)       |
| [AVL Tree](https://en.wikipedia.org/wiki/AVL_tree)                                              | [:octocat:](avl_tree/C)               | [:octocat:](avl_tree/Cpp)             | [:octocat:](avl_tree/Java)            | [:octocat:](avl_tree/Python)          |
| [Binary Search Tree](https://en.wikipedia.org/wiki/Binary_search_tree)                          |                                       | [:octocat:](binary_search_tree/Cpp)   |                                       |                                       |
| [Fenwick Tree](https://en.wikipedia.org/wiki/Fenwick_tree)                                      |                                       | [:octocat:](fenwick_tree/Cpp)         | [:octocat:](fenwick_tree/java)        | [:octocat:](fenwick_tree/Python)      |
| [Doubly Linked List](https://en.wikipedia.org/wiki/Doubly_linked_list)                          | [:octocat:](doubly_linked_list/c)     |                                       |                                       |                                       |
```

File: README.md [hunk 4]
```diff
@@ -12,43 +12,43 @@ This repository contains examples of various algorithms written on different pro
...
-| C               | <pre>gcc [filename.c]<br>./a.out  # unix<br>a.exe  # windows</pre>     |
-| CPP             | <pre>g++ [filename.cpp]<br>./a.out # unix<br>a.exe # windows</pre>     |
-| Java            | <pre>javac [filename.java]<br>java [filename]</pre>                    |
+| Language   | Steps                                                              |
+| ---------- | ------------------------------------------------------------------ |
+| JavaScript | <pre>node [filename.js]</pre>                                      |
+| Python     | <pre>python [filename.py]</pre>                                    |
+| C          | <pre>gcc [filename.c]<br>./a.out # unix<br>a.exe # windows</pre>   |
+| CPP        | <pre>g++ [filename.cpp]<br>./a.out # unix<br>a.exe # windows</pre> |
+| Java       | <pre>javac [filename.java]<br>java [filename]</pre>                |
```
Comment (sangamcse) [inline README.md:51]:
```suggestion
| Language        | Steps                                                                  |
| --------------- | ---------------------------------------------------------------------- |
| JavaScript      | <pre>node [filename.js]</pre>                                          |
| Python          | <pre>python [filename.py]</pre>                                        |
| C               | <pre>gcc [filename.c]<br>./a.out  # unix<br>a.exe  # windows</pre>     |
| CPP             | <pre>g++ [filename.cpp]<br>./a.out # unix<br>a.exe # windows</pre>     |
| Java            | <pre>javac [filename.java]<br>java [filename]</pre>                    |
```

File: README.md [hunk 5]
```diff
@@ -12,43 +12,43 @@ This repository contains examples of various algorithms written on different pro
...
+| [QuickSort](https://en.wikipedia.org/wiki/Quicksort)                               |   [:octocat:](quicksort/C)    |                                 |   [:octocat:](quicksort/Java)   |     [:octocat:](quicksort/Python)     |
+| [Insertion Sort](https://en.wikipedia.org/wiki/Insertion_sort)                     | [:octocat:](insertion_sort/C) | [:octocat:](insertion_sort/Cpp) |                                 |  [:octocat:](insertion_sort/Python)   |
+| [Counting Sort](https://en.wikipedia.org/wiki/Counting_sort)                       | [:octocat:](counting_sort/C)  | [:octocat:](counting_sort/Cpp)  |                                 |   [:octocat:](counting_sort/Python)   |
+| [Radix Sort](https://en.wikipedia.org/wiki/Radix_sort)                             |                               |   [:octocat:](radix_sort/Cpp)   |                                 |    [:octocat:](radix_sort/Python)     |
+| [Binary Search](https://en.wikipedia.org/wiki/Binary_search_algorithm)                          | [:octocat:](binary_search/C) | [:octocat:](binary_search/Cpp)        | [:octocat:](binary_search/C)          | [:octocat:](binary_search/Python)     |
```
Comment (sangamcse) [inline README.md:23]: You have listed you file at 2 places in this line, one in `C` column and other in `Java` column. I hope you understand what you are doing here

File: README.md [hunk 6]
```diff
@@ -12,43 +12,43 @@ This repository contains examples of various algorithms written on different pro
...
+| [Shell Sort](https://en.wikipedia.org/wiki/Shellsort)                              |   [:octocat:](shell_sort/C)   |                                 |                                 |    [:octocat:](shell_sort/Python)     |
+| [Heap Sort](https://en.wikipedia.org/wiki/Heapsort)                                |                               |                                 |                                 |     [:octocat:](heap_sort/python)     |
+| [Maximum Subarray Problem](https://en.wikipedia.org/wiki/Maximum_subarray_problem) |                               |                                 |                                 | [:octocat:](/maximum_subarray/Python) |
+| [Knapsack Problem](https://en.wikipedia.org/wiki/Knapsack_problem)                 |                               |                                 |                                 | [:octocat:](knapsack_problem/Python)  |
+| [Selecton Sort](https://en.wikipedia.org/wiki/Selection_sort)                      | [:octocat:](selection_sort/C) | [:octocat:](selection_sort/Cpp) |                                 |                                       |
```
Comment (sangamcse) [inline README.md:29]:
```suggestion
| Algorithm                                                                                       | C                                     | CPP                                   | Java                                  | Python                                |
|:----------------------------------------------------------------------------------------------- |:-------------------------------------:|:-------------------------------------:|:-------------------------------------:|:-------------------------------------:|
| [Euclidean GCD](https://en.wikipedia.org/wiki/Euclidean_algorithm)                              | [:octocat:](euclidean_gcd/C)          |                                       | [:octocat:](euclidean_gcd/Java)       | [:octocat:](euclidean_gcd/Python)     |
| [Merge Sort](https://en.wikipedia.org/wiki/Merge_sort)                                          | [:octocat:](merge_sort/C)             | [:octocat:](merge_sort/Cpp)           | [:octocat:](merge_sort/Java)          | [:octocat:](merge_sort/Python)        |
| [QuickSort](https://en.wikipedia.org/wiki/Quicksort)                                            | [:octocat:](quicksort/C)              |                                       | [:octocat:](quicksort/Java)           | [:octocat:](quicksort/Python)         |
| [Insertion Sort](https://en.wikipedia.org/wiki/Insertion_sort)                                  | [:octocat:](insertion_sort/C)         | [:octocat:](insertion_sort/Cpp)       |                                       | [:octocat:](insertion_sort/Python)    |
| [Counting Sort](https://en.wikipedia.org/wiki/Counting_sort)                                    | [:octocat:](counting_sort/C)          | [:octocat:](counting_sort/Cpp)        |                                       | [:octocat:](counting_sort/Python)     |
| [Radix Sort](https://en.wikipedia.org/wiki/Radix_sort)                                          |                                       | [:octocat:](radix_sort/Cpp)           |                                       | [:octocat:](radix_sort/Python)        |
| [Binary Search](https://en.wikipedia.org/wiki/Binary_search_algorithm)                          | [:octocat:](binary_search/C)          | [:octocat:](binary_search/Cpp)        |                                       | [:octocat:](binary_search/Python)     |
| [Bubble Sort](https://en.wikipedia.org/wiki/Bubble_sort)                                        | [:octocat:](bubble_sort/C)            | [:octocat:](bubble_sort/Cpp)          | [:octocat:](bubble_sort/Java)         | [:octocat:](bubble_sort/Python)       |
| [Shell Sort](https://en.wikipedia.org/wiki/Shellsort)                                           | [:octocat:](shell_sort/C)             |                                       |                                       | [:octocat:](shell_sort/Python)        |
| [Heap Sort](https://en.wikipedia.org/wiki/Heapsort)                                             |                                       |                                       |                                       | [:octocat:](heap_sort/python)         |
| [Maximum Subarray Problem](https://en.wikipedia.org/wiki/Maximum_subarray_problem)              |                                       |                                       |                                       | [:octocat:](/maximum_subarray/Python) |
| [Knapsack Problem](https://en.wikipedia.org/wiki/Knapsack_problem)                              |                                       |                                       |                                       | [:octocat:](knapsack_problem/Python)  |
| [Selecton Sort](https://en.wikipedia.org/wiki/Selection_sort)                                   | [:octocat:](selection_sort/C)         | [:octocat:](selection_sort/Cpp)       |                                       |                                       |
```

**Review by sangamcse — CHANGES_REQUESTED** [review]
I have added few suggestions that needs to be done and make sure to do **`git commit --amend`** instead of `git commit`. And in the commit message, add line break after 80 characters so that each line just consists 80 chars. For ref, https://github.com/NITSkmOS/Algorithms/blob/master/CONTRIBUTING.md#commit-body.

Rest looks good to me.

**Pull request closed without merging**